server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -541,19 +578,79 @@ private static void addGeneratedToVerify(Map<ColumnIdent, ColumnConstraint> colu
         }
     }
 
-    @SuppressWarnings("unchecked")
-    public ParsedDocument index(IndexItem item) throws IOException {
-        assert item.insertValues().length <= valueIndexers.size()
-            : "Number of values must be less than or equal the number of targetColumns/valueIndexers";
+    /**
+     * Indicates whether inserts should use OID in source.
+     * @return true if:
+     * <ol>
+     * <li>Table is created on Version on or after 5.5, ie we are not inserting into an old table, restored from a snapshot.</li>
+     * <li>All cluster nodes support updating schema with getting OIDs in response.</li>
+     * </ol>
+     */
+    public boolean targetsHaveOids() {
+        return targetsHaveOids;
+    }
```

Comment:
```suggestion
```

Could have something like this instead in the TransportShardUpsertAction:

```
        if (response instanceof AddColumnResponse addColumnResponse) {
            indexer.updateTargets(addColumnResponse.addedColumns());
        }

```

"User needs to call `updateTargets` after columns from `collectSchemaUpdates` were added to the cluster state" is a simpler contract than making it aware of the details.


But see the comment further above about the `targetsHaveOids` condition. I think we'll have to change the approach anyway.

Reply:
Thanks, applied in https://github.com/crate/crate/pull/14617/commits/287daade95494977a3ff35f6ff6c80eb4f75ecd7